### ISSUE TITLE
Move online data folder from OSF to Zenodo

### ISSUE THREAD

**Issue by stefsmeets** [FormingWorlds/PROTEUS#242]
Hi all, #220 changes something and now the dummy input fails with:
`FileNotFoundError: /home/stef/python/proteus/fwl_data/mass_radius/Mass-radius/Zeng2019/massradiusEarthlikeRocky.txt not found.`

I seem to be missing some file, but it is unclear to me how I can get it.

**Comment by stefsmeets** [comment]
Seems like I need to run `proteus get reference`, but it says that the data already exists:

```console
$ find . -name '*massradius*'
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradius_50percentH2O_1000K_1mbar.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradius_50percentH2O_500K_1mbar.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradius_100percentH2O_500K_1mbar.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradiusFe.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradius_50percentH2O_300K_1mbar.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradius_100percentH2O_300K_1mbar.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradius_100percentH2O_700K_1mbar.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradiushydrogen.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradiusMaxCollisionStripping.txt
./fwl_data/mass_radius/Mass-radius/Zeng2019/massradiusmgsio3.txt
```

**Comment by stefsmeets** [comment]
I removed the mass_radius directory and ran `proteus get reference` again. Seems like the download command only checks for the presence of the directory. This seems like an oversight, especially because it fails silently.

**Comment by nichollsh** [comment]
Is this true for all of the download commands then? Or just this one?

**Comment by stefsmeets** [comment]
I don't know, I didn't look at the other ones.

**Comment by stefsmeets** [comment]
Oof, there is a lot of room for refactoring the `proteus.utils.data` module. All download functions check for the existance of the directory, and skips downloading if the directory exists. If any of the files inside the directory are updated, then these will not be downloaded.

However, checking for every file is also extremely slow. And I believe OSF does not have something like a file index to check against, you must check file-by-file which is a performance killer.

**Comment by nichollsh** [comment]
It might be possible to get this information using the API?
e.g. `curl -X "GET" https://api.osf.io/v2/nodes/fzwr4/files/osfstorage/`

**Comment by nichollsh** [comment]
Actually, this can be done with the osf library that we are already using: https://osfclient.readthedocs.io/en/latest/api/osfclient.html#osfclient.cli.list_

**Comment by nichollsh** [comment]
And one other thing: while working on this, it might be a good time to address #206.

**Comment by stefsmeets** [comment]
@lsoucasse I don't believe this was fixed. #206 is not a duplicate of this issue.

**Comment by lsoucasse** [comment]
> @lsoucasse I don't believe this was fixed. #206 is not a duplicate of this issue.

Sorry, thought PR #244 fully addressed it.

**Comment by nichollsh** [comment]
There are still ongoing issues with obtaining data from OSF. Sometimes it manifests as missing files (as in this issue) and sometimes just as an HTTP 502 error.

**Comment by timlichtenberg** [comment]
Ok, so this is probably connected the discussion we had on moving things to Zenodo. I cannot find the issue for that, did we make one?

**Comment by nichollsh** [comment]
We didn't make an issue for that, but yeah it's definitely connected. Are we set on Zenodo then?

**Comment by timlichtenberg** [comment]
But I distinctly remember somehow that we talked about specific links in here somewhere. I eveb set up a Zenodo "community" for testing this out. I think Zenodo makes most sense ultimately. Shall we set up a Forming Worlds or PROTEUS community there slowly?

**Comment by nichollsh** [comment]
We discussed it here: https://github.com/orgs/FormingWorlds/discussions/338

**Comment by nichollsh** [comment]
But yes I think we should at least try it, and see how it goes. There's the downside of not being able to modify the files so easily, but in theory we shouldn't have to do this too much.

**Comment by nichollsh** [comment]
Potential alternative: https://www.cloudflare.com/en-gb/application-services/products/cdn/
This platform is better designed for providing files on demand, compared to both Zenodo and OSF.

They have a free plan - not sure what the 'catch' is.

**Comment by nichollsh** [comment]
Zenodo communities:
https://zenodo.org/communities/forming_worlds
https://zenodo.org/communities/proteus_framework

@timlichtenberg I assume we should upload data to the proteus community?

**Comment by timlichtenberg** [comment]
Yes, I just accepted the request. What's your Zenodo email or username? I can add people as managers to the Zenodo community.

**Comment by nichollsh** [comment]
My username is 'nichollsh'

**Comment by timlichtenberg** [comment]
I cannot find that account. Maybe it's because of this:
> only users with a public profile can be invited

**Comment by nichollsh** [comment]
Ah, apparently my account was hidden. Can you try now?

**Comment by timlichtenberg** [comment]
Yep, works now, sent you an invite.

**Comment by nichollsh** [comment]
Thanks! Seems to work.

We should think about how to store data on the community. How coarsely do we define the entries to be? 

For example, does a single spectral file get its own record? Or do we put spectral files into a single record, and then update it over time?

**Comment by nichollsh** [comment]
Also, since the records are technically independent entities, it might be worth setting out some standards for how to name/tag them.

**Comment by timlichtenberg** [comment]
For simplicity we could recreate the OSF structure. However, I think that our PROTEUS organisation has evolved a bit since then. Would it make sense to match the sub-module structure from within PROTEUS then for example? It does not make sense to label according to sub-modules, as several modules will use the same input data.

**Comment by nichollsh** [comment]
Following the module structure (`interior`, `atmos_clim`, `outgas`, etc) makes a lot of sense.

We will probably still need >1 record per module though, as each one could easily have >100 files.

Maybe prefixing record names with the module would be a good way to handle this?
Examples:
* The record `fwl-interior-eos`: interior EOS data
* The record `fwl-star-spectra`: stellar spectra
* The record `fwl-atmos_clim-honeyside`: honeyside spectral file

**Comment by nichollsh** [comment]
We have decided to create 1 record per module for now. 
Ideally the 50 GB size limit should not be an issue - we should avoid having this much data.
The 100 file limit per record can be mitigated by zipping the files.

Proposed Zenodo records below. Please edit this comment if required.

### tests
test_data

### atmos_clim
spectral_files
surface_albedos
thermodynamics
parfiles 

### star
stellar_spectra
tracks

### interior
melting_tables
eos_tables

### atmos_chem
networks

### struct
?

### observe
?

### examples
example_outputs

**Comment by timlichtenberg** [comment]
Looks good to me. My question from earlier is if we need the "struct" part, as Zalmoxis at least will and optimally should use the EOS that also Aragog uses, so the tables would be duplicate. There may be some non-overlapping cases though, so maybe these need to be split up. @flaviacris03 @planetmariana, your thoughts?

**Comment by lsoucasse** [comment]
I have just checked online the data we have on Zenodo. I see the surface albedo from Hammond. Are there other data that are private @nichollsh (I am not a member)? Anyway, I will start playing with this.

**Comment by nichollsh** [comment]
I haven't added any other data. These were just uploaded as a test, and don't really fit into the structure that I proposed above. Thanks for looking into this.

Here's a link the the Zenodo Community page, for those who need it: https://zenodo.org/communities/proteus_framework/records?q&l=list&p=1&s=20&sort=newest

**Comment by nichollsh** [comment]
So rather than having 1 Zenodo Record per module, containing several zip files, we would have 1 record per 'dataset'. This would mean 1 Record for each of the target items currently in `data.py`? 

E.g. there would be a record for the exoplanet reference data (OSF fzwr4) and one record per spectral file (.sf + .sf_k)?

**Comment by lsoucasse** [comment]
> So rather than having 1 Zenodo Record per module, containing several zip files, we would have 1 record per 'dataset'. This would mean 1 Record for each of the target items currently in `data.py`?
> 
> E.g. there would be a record for the exoplanet reference data (OSF fzwr4) and one record per spectral file (.sf + .sf_k)?

I would say one record for each OSF "final" folder, like `spectral_files/Honeyside/256`

**Comment by lsoucasse** [comment]
My Zenodo username is Laurent_Soucasse.

**Comment by lsoucasse** [comment]
> > Harrsion, what was your argument against many different Zenodo repositories again?
> 
> My main concern relates to how data are structured on Zenodo. Unlike OSF, we cannot structure the data hierarchically, so it will not necessarily be clear how different records relate to each other. We might be able to mitigate this by naming them clearly and in a standard manner. There's also the option of adding tags.
> 
> E.g. the spectral file Honeyside48 could have a record named: `fwl/atmos/spfile/Honeyside48` so that it appears alongside other `atmos/` records when listed alphabetically. There's no way to nest them, so it would not be possible to contain all spectral files in a 'folder' of some kind. However, this does mean that the name for each record is somewhat confusing and not so human-readable. I think this is what Laurent might be suggesting above?
> 
> E.g. the record for Spada tracks could be named `fwl/star/tracks/Spada13`, and the Baraffe one named `fwl/star/tracks/Baraffe15`.
> 
> We will need to be rigorous in ensuring that records have consistent names and descriptions.

That was my idea of naming the Zenodo records. But indeed, it means we are rigorous and consistent with the OSF and this cannot really be automated.
